proposals/000-content-iterator.md

```diff
@@ -0,0 +1,223 @@
+
+:warning: Don't forget to `close` the iterator when you are done, to discard opened resources.
+
+### Extracting the data from `Content` elements
+
+The `Content` elements are value objects containing:
+
+* a `Locator` targeting the piece of content
+* the associated `Data`
+
+There are several types of `Data` which can be returned by the iterator. Depending on your use case, you might want to filter on the type of data to get only what you need.
+
+#### Media data
+
+Media data are returned when a resource embeds another media resource in the reading flow. They hold a publication `Link` to the embedded resource.
+
+Two kind of media data are currently supported:
+
+* `Audio` for audio clips.
+* `Image` for embedded images.
+    * It also holds an optional `description` string for accessibility purposes.
+
+#### Text data
+
+The `Text` data is used for the text elements inlined in the publication resources. Each text element matches a semantic item represented as a `TextStyle` in the data object:
+
+* `heading(level: Int)` for text headings, with an associated level
+* `body` for a basic body paragraph
+* `caption` for a caption associated to an image
+* `footnote` for footnotes at the end of the resource
+* `quote` for a blockquote
+* `listItem` for a single list item
+
+Each `Text` data is split in one or more spans containing the text content, a `Locator` and the language. 
+
+## Reference Guide
+
+### Types and APIs
+
+#### `ContentIterationService` Interface (implements `PublicationService`)
+
+Provides `ContentIterator` instances to crawl the content of a `Publication`.
+
+##### Methods
+
+* `iteratorFrom(start: Locator?) -> ContentIterator?`
+    * Creates a `ContentIterator` starting from the given location.
+    * Returns `null` if no iterator can be created, for example because no resources are iterable.
+
+##### `Publication` Helpers
+
+* `isContentIterable: Boolean = findService(ContentIterationService::class) != null`
+    * Returns whether this `Publication` can be iterated on.
+* `contentIteratorFrom(start: Locator?) -> ContentIterator? = findService(ContentIterationService::class)?.iteratorFrom(start)`
+    * Creates a `ContentIterator` starting from the given location.
+
+#### `ContentIterator` Interface
+
+Iterates over `Content` elements.
+
+This interface does not depend on `Publication`.
+
+##### Methods
+
+* `previous() -> Content?`
+    * Returns the previous `Content` element in the iterator, or `null` when reaching the beginning.
+* `next() -> Content?`
+    * Returns the next `Content` element in the iterator, or `null` when reaching the end.
+* `close()`
+    * Closes and discard resources held by this iterator.
+
+#### `Content` Class
+
+Represents a single semantic content element.
+
+##### Properties
+
+* `locator: Locator`
+    * Locator targeting this element in the `Publication`.
```

Comment:
I can't guarantee that it will be the case for all media types, but for the ones we have so far I think so.

With image elements, you can use a `cssSelector` in the `locations` object. I filled it for all locators in the [`HtmlResourceContentIterator`](https://github.com/readium/swift-toolkit/blob/f074a933a4dd81ae0be0b22373fa80af7af4d3af/Sources/Shared/Publication/Services/Content/HTMLResourceContentIterator.swift#L137), as even with a text context it can help limit the search scope to  a parent element.

If there's ever a case where we can't target precisely, this `Locator` should at least match the closest targetable parent.

Reply:
Ok, I see.